cni/network/network.go

```diff
@@ -453,274 +527,221 @@ func (plugin *NetPlugin) Add(args *cniSkel.CmdArgs) error {
+		// TODO figure out where to put telemetry: sendEvent(plugin, fmt.Sprintf("CNI ADD succeeded: IP:%+v, VlanID: %v, podname %v, namespace %v numendpoints:%d",
+		//	ipamAddResult.interfaceInfo[ifIndex].IPConfigs, epInfo.Data[network.VlanIDKey], k8sPodName, k8sNamespace, plugin.nm.GetNumberOfEndpoints("", nwCfg.Name)))
+		i++
 	}
-
-	nwDNSInfo, err := getNetworkDNSSettings(ipamAddConfig.nwCfg, ipamAddResult.defaultInterfaceInfo.DNS)
+	cnsclient, err := cnscli.New(nwCfg.CNSUrl, defaultRequestTimeout)
 	if err != nil {
-		err = plugin.Errorf("Failed to getDNSSettings: %v", err)
-		return nwInfo, err
-	}
-
-	logger.Info("DNS Info", zap.Any("info", nwDNSInfo))
-
-	// Create the network.
-	nwInfo = network.NetworkInfo{
-		Id:                            networkID,
-		Mode:                          ipamAddConfig.nwCfg.Mode,
-		MasterIfName:                  masterIfName,
-		AdapterName:                   ipamAddConfig.nwCfg.AdapterName,
-		BridgeName:                    ipamAddConfig.nwCfg.Bridge,
-		EnableSnatOnHost:              ipamAddConfig.nwCfg.EnableSnatOnHost,
-		DNS:                           nwDNSInfo,
-		Policies:                      policies,
-		NetNs:                         ipamAddConfig.args.Netns,
-		Options:                       ipamAddConfig.options,
-		DisableHairpinOnHostInterface: ipamAddConfig.nwCfg.DisableHairpinOnHostInterface,
-		IPV6Mode:                      ipamAddConfig.nwCfg.IPV6Mode, // TODO: check if IPV6Mode field can be deprecated
-		IPAMType:                      ipamAddConfig.nwCfg.IPAM.Type,
-		ServiceCidrs:                  ipamAddConfig.nwCfg.ServiceCidrs,
-		IsIPv6Enabled:                 ipamAddResult.ipv6Enabled,
-	}
-
-	if err = addSubnetToNetworkInfo(ipamAddResult, &nwInfo); err != nil {
-		logger.Info("Failed to add subnets to networkInfo",
-			zap.Error(err))
-		return nwInfo, err
+		return errors.Wrap(err, "failed to create cns client")
 	}
-	setNetworkOptions(ipamAddResult.ncResponse, &nwInfo)
+	defer func() {
+		if err != nil {
 
-	err = plugin.nm.CreateNetwork(&nwInfo)
+			// Delete all endpoints
+			for _, epInfo := range epInfos {
+				deleteErr := plugin.nm.DeleteEndpoint(epInfo.NetworkID, epInfo.EndpointID, epInfo)
+				if deleteErr != nil {
+					// we already do not return an error when the endpoint is not found, so deleteErr is a real error
+					logger.Error("Could not delete endpoint after detecting add failure", zap.String("epInfo", epInfo.PrettyString()), zap.Error(deleteErr))
+					return
+				}
+			}
+			// Rely on cleanupAllocationOnError declared above to delete ips
+			// Delete state in disk here
+			delErr := plugin.nm.DeleteState(epInfos)
+			if delErr != nil {
+				logger.Error("Could not delete state after detecting add failure", zap.Error(delErr))
+				return
+			}
+		}
+	}()
+	// CHECK: why does the cns client deal with apipa?-- "cns client" only used in windows to create apipa
+	err = plugin.nm.EndpointCreate(cnsclient, epInfos)
```

Comment:
can we remove `cns.BackendNIC:` case since its noop anyway and can be matched with default

Reply:
fix linter issue